### PR TITLE
Shutdown the executor gracefully on scale-down.

### PR DESCRIPTION
**Pull request by tylerwilliams**
Ensure in-flight work is cancelled before the shutdown and no more messages are sent.

### CODE REVIEW

File: enterprise/server/scheduling/priority_task_scheduler/priority_task_scheduler.go [hunk 1]
```diff
@@ -157,41 +158,45 @@ func NewPriorityTaskScheduler(env environment.Env, exec *executor.Executor, opti
...
-			qes.mu.Lock()
-			activeTasks := len(qes.activeTaskCancelFuncs)
-			qes.mu.Unlock()
-			if activeTasks == 0 {
-				break
-			}
-			time.Sleep(100 * time.Millisecond)
+		deadline, ok := ctx.Deadline()
+		if !ok {
+			rootCancel()
```
Comment (vadimberezniker) [inline enterprise/server/scheduling/priority_task_scheduler/priority_task_scheduler.go:189]: Would this case ever be hit?

File: enterprise/server/scheduling/priority_task_scheduler/priority_task_scheduler.go [hunk 2]
```diff
@@ -157,41 +158,45 @@ func NewPriorityTaskScheduler(env environment.Env, exec *executor.Executor, opti
...
-	// but not started yet, ensuring that someone else will have a chance to get it.
+
+	// This func ensures that we don't attempt to claim work that was
+	// enqueued but not started yet, allowing another executor a chance
+	// to complete it. This is client-side "graceful" stop -- we stop
+	// processing queued work as soon as we receive a shutdown signal, but
+	// permit in-progress work to continue, up until just before the
+	// shutdown timeout, at which point we hard-cancel it.
 	env.GetHealthChecker().RegisterShutdownFunction(func(ctx context.Context) error {
 		qes.mu.Lock()
```
Comment (vadimberezniker) [inline enterprise/server/scheduling/priority_task_scheduler/priority_task_scheduler.go:182]: It looks like the lock is being hold while waiting for the shutdown deadline.

**Review by vadimberezniker — CHANGES_REQUESTED** [review]

File: enterprise/server/scheduling/priority_task_scheduler/priority_task_scheduler.go (hunk 2 above)
Comment (tylerwilliams) [reply enterprise/server/scheduling/priority_task_scheduler/priority_task_scheduler.go:183]: my bad; dashed this off before lunch without testing first oO

**Review by vadimberezniker — APPROVED** [review]
LGTM

**Review by vadimberezniker — APPROVED** [review]

**Pull request merged**